File: Populating_Next_Right_Pointers_in_Each_Node.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2014 fuel <fuel@fuel-3330>
#
# Distributed under terms of the MIT license.

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x, left=None, right=None):
        self.val = x
        self.left = left
        self.right = right
        self.next = None
    def pl(self):
        if(self.next):
            print self.val,"->",
            self.next.pl()
        else:
            print self.val

class Solution:
    # @param root, a tree node
    # @return nothing
    def connect(self, root):
        def connect2(ln,rn):
            ln.next=rn
            return rn
        if root==None:
            return
        prelevel = [root]
        while prelevel!=[]:
            reduce(connect2,prelevel)
            curlevel = []
            for node in prelevel:
                if node.left!=None: curlevel.append(node.left)
                if node.right!=None: curlevel.append(node.right)
            prelevel = curlevel

T = TreeNode
s = Solution()
tree = T(0,
         T(1, T(3), T(4)),
         T(2, T(5), T(6)))
s.connect(tree)

def connect2(ln,rn):
    ln.next=rn
    return rn
#
# nodes = [T(1),T(2),T(3)]
# nodes = [T(1),T(2)]
# reduce(connect2,nodes)
# print [n.val for n in nodes]
# head = nodes[0]
# head.pl()

File: Algorithms/Populating_Next_Right_Pointers_in_Each_Node.cpp
/*
 * Binary_Tree_Level_Order_Traversal.cpp
 * Copyright (C) 2014 fuel <fuel@fuel-3330>
 *
 * Distributed under terms of the MIT license.
 */

#include <iostream>
#include <vector>
using namespace std;
// Definition for binary tree
struct TreeLinkNode {
    int val;
    TreeLinkNode *left;
    TreeLinkNode *right;
    TreeLinkNode *next;
    TreeLinkNode(int x, TreeLinkNode* left=nullptr, TreeLinkNode* right=nullptr) : val(x), left(left), right(right) {}
    void pl() {
        if(this->next==nullptr) cout<<this->val<<endl;
        else {
            cout<<this->val<<"->";
            this->next->pl();
        }
    }
};

class Solution {
public:
    void connect(TreeLinkNode* p) {
        if (p == NULL || p->left == NULL || p->right == NULL)
            return;
        TreeLinkNode* rightSibling;
        TreeLinkNode* p1 = p;
        while (p1) {
            if (p1->next)
                rightSibling = p1->next->left;
            else
                rightSibling = NULL;
            p1->left->next = p1->right;
            p1->right->next = rightSibling;
            p1 = p1->next;

        }
        connect(p->left);
    }
    vector<vector<int> > levelOrder(TreeLinkNode *root) { 
        if(root==nullptr) return {};
        vector<vector<int> > res;
        vector<TreeLinkNode*> *curlevel = new vector<TreeLinkNode*>{root};
        while(curlevel->size()) {
            (*curlevel)[0]->pl();
            res.push_back({});
            vector<int> &curvals = res[res.size()-1];
            vector<TreeLinkNode*> *nextlevel = new vector<TreeLinkNode*>{};
            for(auto p : *curlevel) {
                curvals.push_back(p->val);
                if(p->left!=nullptr) nextlevel->push_back(p->left);
                if(p->right!=nullptr) nextlevel->push_back(p->right);
            }
            delete curlevel;
            curlevel=nextlevel;
        }
        return res;
    }
};
int main() {
    #define N new TreeLinkNode
    TreeLinkNode *tree =  N(0,
                            N(1,N(3),N(4)),
                            N(2,N(5),N(6)));
    Solution s;
    s.connect(tree);
    cout<<"[\n";
    for (auto level : s.levelOrder(tree)) {
        cout<<"  [";
        for(auto e : level) 
            cout<<e<<" ";
        cout<<"]\n";
    }
    cout<<"]\n";

    return 0;
}
